packages/autoloader/src/class-version-selector.php

```diff
@@ -0,0 +1,92 @@
+<?php
+/* HEADER */ // phpcs:ignore
+
+/**
+ * Used to select package versions.
+ */
+class Version_Selector {
+
+	/**
+	 * Indicates whether a package class or file's plugin has a development version.
+	 *
+	 * @var Array
+	 */
+	private static $dev_plugin_in_use = array();
+
+	/**
+	 * Checks whether the selected package version should be updated.
+	 *
+	 * @param String $class The class name.
+	 * @param String $selected_version The currently selected package version.
+	 * @param String $compare_version The package version that is being evaluated to
+	 *                                determine if the version needs to be updated.
+	 * @param String $plugin The name of the plugin that contains the package.
+	 *
+	 * @return Boolean Returns true if the selected package version should be updated,
+	 *                 else false.
+	 */
+	public function update_package_version( $class, $selected_version, $compare_version, $plugin ) {
+
+		if ( is_null( $selected_version ) ) {
+			if ( $this->is_plugin_version_dev( $plugin ) ) {
+				self::$dev_plugin_in_use[ $class ] = true;
+			}
+
+			return true;
+		}
+
+		if ( $this->is_package_version_dev( $selected_version ) && self::$dev_plugin_in_use[ $class ] ) {
+			return false;
+		}
+
+		if ( $this->is_package_version_dev( $compare_version ) && $this->is_plugin_version_dev( $plugin ) ) {
+			self::$dev_plugin_in_use[ $class ] = true;
+			return true;
+		}
+
+		if ( version_compare( $selected_version, $compare_version, '<' ) ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Checks whether the given package version is a development version.
+	 *
+	 * @param String $version The package version.
+	 *
+	 * @return Boolean True if the version is a dev version, else false.
+	 */
+	private function is_package_version_dev( $version ) {
+		if ( 'dev-' === substr( $version, 0, 4 ) || '9999999-dev' === $version ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Checks whether the given plugin has a development version.
+	 *
+	 * @param String $plugin The plugin's slug.
+	 *
+	 * @return Boolean True if the plugin version is a dev version, else false.
+	 */
+	private function is_plugin_version_dev( $plugin ) {
+		$dir         = explode( '/', $plugin )[0];
+		$plugin_data = get_plugins( "/$dir" );
```

Comment:
What do you say we solely rely here on `WP_DEBUG` and/or `JETPACK__DEV_DEBUG` constants? I think that would make everyone's life a lot easier. Want dev autoloader versions - turn on `WP_DEBUG` :)

Reply:
This is great idea, and I think that would definitely make everyone's life easier! 

I think that I'd prefer to use `JETPACK__DEV_DEBUG` alone, since production sites occasionally have `WP_DEBUG` set to true, and it seems less likely that they would have `JETPACK__DEV_DEBUG` set to true.

Reply:
One thing that we need to keep in mind here is that `JETPACK__DEV_DEBUG` also turns off a lot of stuff, I didn't think about that right away. Using that constant will mean that we won't be able to test some things.

Reply:
Good point; I didn't know about that.

Do you agree that we shouldn't use `WP_DEBUG` because some production sites set it to true? What do you think of using a new constant, something like `JETPACK_AUTOLOADER_DEV_DEBUG`?

Reply:
Yes, I agree about `WP_DEBUG` and yes, I wanted to suggest the same thing - let's introduce a new constant, but let's make it a bit easier on the eyes, something like `JETPACK_AUTOLOAD_DEV`? Or maybe you can come up with a better/shorter one? :)

Reply:
I couldn't think of a short but still descriptive name, so I just went with `JETPACK_AUTOLOAD_DEV` in commit ea25283